posthog/temporal/batch_exports/pre_export_stage.py

```diff
@@ -0,0 +1,561 @@
+    if is_5_min_batch_export(full_range=full_range):
+        clickhouse_url = settings.CLICKHOUSE_OFFLINE_5MIN_CLUSTER_HOST
+
+    # Data can sometimes take a while to settle, so for 5 min batch exports we wait several seconds just to be safe.
+    # For all other batch exports we wait for 1 minute since we're querying the events_recent table using a
+    # distributed table and setting `max_replica_delay_for_distributed_queries` to 1 minute
+    if is_5_min_batch_export(full_range):
+        delta = dt.timedelta(seconds=30)
+    else:
+        delta = dt.timedelta(minutes=1)
+    end_at = full_range[1]
+    await wait_for_delta_past_data_interval_end(end_at, delta)
+
+    done_ranges = []
+    async with get_client(team_id=team_id, clickhouse_url=clickhouse_url) as client:
+        if not await client.is_alive():
+            raise ConnectionError("Cannot establish connection to ClickHouse")
+
+        # TODO - in future we might want to catch any ClickHouse memory usage errors and break down the interval into
+        # sub-intervals to reduce memory usage
+        for interval_start, interval_end in generate_query_ranges(full_range, done_ranges):
+            if interval_start is not None:
+                query_parameters["interval_start"] = interval_start.strftime("%Y-%m-%d %H:%M:%S.%f")
+            query_parameters["interval_end"] = interval_end.strftime("%Y-%m-%d %H:%M:%S.%f")
+            query_id = uuid.uuid4()
+
+            if isinstance(query_or_model, RecordBatchModel):
+                query, query_parameters = await query_or_model.as_query_with_parameters(interval_start, interval_end)
+            else:
+                query = query_or_model
+
+            s3_staging_folder = _get_s3_staging_folder(
+                batch_export_id=batch_export_id,
+                data_interval_start=data_interval_start,
+                data_interval_end=data_interval_end,
+            )
+            try:
+                await _delete_all_from_bucket_with_prefix(
+                    bucket_name=settings.BATCH_EXPORT_INTERNAL_STAGING_BUCKET, key_prefix=s3_staging_folder
+                )
+            except Exception as e:
+                await logger.aexception("Unexpected error occurred while deleting existing objects from S3", exc_info=e)
+                raise
+
+            try:
+                await client.execute_query(query, query_parameters=query_parameters, query_id=str(query_id))
+                # TODO - remove this once testing over
+                # need to wait for query info to become available in system.query_log
+                await asyncio.sleep(5)
+                memory_usage = await client.read_query(
+                    f"SELECT formatReadableSize(memory_usage) as memory_used FROM system.query_log WHERE query_id = '{query_id}' AND type='QueryFinish' ORDER BY event_time DESC LIMIT 1",
+                )
+                await logger.ainfo(f"Query memory usage = {memory_usage.decode('utf-8').strip()}")
```

Comment:
Maybe consider putting this in a background task instead so as to not block the main thread? I know we are testing but still: What if it takes longer than 5 seconds?